_published/0013-authentication.md

```diff
@@ -0,0 +1,165 @@
+built in a minimalist fashion that offloads most of the work involved in
+administration to other pre-existing platforms.
+
+![Connectivity visualised](../assets/auth-map-v2.PNG)
+
+#### The ID Token
+A OIDC type ID-token can be used for authentication. The custom claims
+(requires an IDP with support for RBAC) specified within the ID-token can
+then be used to specify access levels. To keep the scope manageable the task
+of dividing up and refining access levels and permission will be left to a
+later date. For this RFC all kinds of read permissions won't require valid
+ID-tokens, but all write type requests will require valid ID-tokens that
+have been signed by a pre-approved issuer.
+The OIDC protocol defines a token called an
+[ID Token](https://auth0.com/docs/tokens/id-tokens/id-token-structure).
+The ID token is signed by the source issuer. This signature is then used to
+validity and authenticity of the token. See
+[validating an OIDC ID-Token](https://connect2id.com/blog/how-to-validate-an-openid-connect-id-token).
+
+#### Sessions
+The wharf-api is currently stateless and does not use sessions. It is
+possible to either include the ID token in every request to the server or to
+start using sessions to store the ID token between requests.
+
+#### Local development vs production environments
+For the initial release of the feature the configuration will be tailored to
+using azure as an IDP. These settings will as necessary be exposed as
+environment variables to each respective docker image.
+
+
+#### Choice of Client
+
+The OIDC standard defines a
+[list of certified clients](https://openid.net/developers/certified/)
+that can help simplify the implementation of ID Token retrieval.
+
+For the backend a full client will not be required. As the tokens just need
+to be validated and not requested. Validation can be done offline by caching
+the issuers' signature for future validations (OBS! The signature key may
+rotate on a timely basis).
+
+## Compatibility
+
+Expect no compatability between versions of wharf without authentication to
+versions of wharf with authentication. There is a potential threat in leaving
+the system backwards compatible in that a non-secured version of the api could
+run together in an otherwise secure cluster.
+
+We should aim to pick a solution where any IDP supporting OIDC and RBAC can be
+used. Locking ourselves to a single IDP when a standard such as OIDC exists
+seems like a  waste opportunity. Keeping the choice of IDP open will leave
+more room for choice and flexibility in the future.
+
+## Alternative solutions
+
+There are many types of identity verification systems.
+This RFC assumes that OIDC will be used. Within OIDC there are multiple
+different [flow definitions](https://developer.okta.com/docs/concepts/oauth-openid/#choosing-an-oauth-2-0-flow).
+We should primarily use `codeflow with pkce` as this is presumed to be the
+most secure, and best compatible at the time of writing.
```

Comment:
I believe the thought was current security practices. Implicit and hybrid code flows dont usually seem to be held in as high a regard for user agent authorisation. 
There is still some intresting testing to be done here for implicit trust towards implicitly trusted clients operating autonomously. There should probably be some info here on how id tokens and custom claims work. That also requires more testing.

Reply:
The limitations here dont seem to be technical as much as beurocratic. On the technical side this is meant to enable even the most complex and secure flows making it easier to convince the beurocrats the tech is ok to implement.
I ran the tests on creating the tokens autonomously as a trusted client and it appears to work in most fashions. Once the tokens are recieved the flow towards the api is identical to the web. The custom roles appear to work as well but production environment final tests are still blocked.

Reply:
The new rewording is slightly more abiguous to allow this to be solved in implementation/configuration rather than stated in the RFC.

Reply:
> making it easier to convince the beurocrats the tech is ok to implement.

Lol

> The new rewording is slightly more abiguous to allow this to be solved in implementation/configuration rather than stated in the RFC.

:+1: 

I'm glad with the explanation. Resolving.